From f2557ab57285c1d6aedcbe5f409ce34b23e5a0a9 Mon Sep 17 00:00:00 2001
From: Lorenzo Bianconi <lorenzo.bianconi@redhat.com>
Date: Thu, 14 May 2026 18:46:01 +0200
Subject: [PATCH] northd: Use MC_UNKNOWN for broadcast ARP requests.

In the OVN logical switch pipeline, broadcast ARP requests (and ND_NS)
generated by VIFs and by router ports are also flooded into the per switch
MC_FLOOD_L2 multicast group which includes all non-router ports of the
switch.
In deployments with large logical broadcast domains (logical switches with
a significantly large number of switch ports, e.g. 200+) this becomes a
problem because the MC_FLOOD_L2 multicast group has a lot of ports so the
chain of the OpenFlow tables that the packet needs to traverse for full
processing becomes really long, going over OVS' 4K resubmit limit and
causing the packet to be dropped.
The main reason why the ARP packets are currently sent to all non-router
ports is to allow the workloads to learn the mapping between arp.spa
(source IP) and arp.sha (source MAC). However, that's just an optimization,
which might avoid future ARP requests from other workloads; and it's not a
requirement, nothing would really break if we didn't forward those packets
to all other VIFs.
It's probably acceptable to change the behavior and just forward these ARP
requests to the ports that have LSP.addresses="unknown" as the workloads
behind those ports (or the fabric for the localnet case) might actually own
the arp.tpa target IP.

Reported-at: https://redhat.atlassian.net/browse/FDP-3439
Co-authored-by: Dumitru Ceara <dceara@redhat.com>
Signed-off-by: Dumitru Ceara <dceara@redhat.com>
Signed-off-by: Lorenzo Bianconi <lorenzo.bianconi@redhat.com>
Signed-off-by: Ales Musil <amusil@redhat.com>
(cherry picked from commit 1872d61e34aad38ad5c5a306b969ce8a211416d8)
Adapted for OVN 25.03 by replacing vector_len(&od->router_ports) with od->n_router_ports.
This matches ovn-org/ovn branch-26.03 8aa51920dda9bc59b05a172bdec55986b8ba2aeb.

diff --git a/northd/northd.c b/northd/northd.c
index 36dae91dce..a7203d9c82 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -8695,6 +8695,7 @@ build_lswitch_rport_arp_req_flow(
 {
     struct ds match   = DS_EMPTY_INITIALIZER;
     struct ds m       = DS_EMPTY_INITIALIZER;
+    struct ds m_garp  = DS_EMPTY_INITIALIZER;
     struct ds actions = DS_EMPTY_INITIALIZER;

     arp_nd_ns_match(ips, addr_family, &m);
@@ -8716,17 +8717,28 @@ build_lswitch_rport_arp_req_flow(
                       patch_op->cr_port->json_key);
     }

+    if (addr_family == AF_INET) {
+        ds_clone(&m_garp, &m);
+        ds_put_format(&m_garp, " && arp.spa == %s", ips);
+    }
+
     /* Send a the packet to the router pipeline.  If the switch has non-router
      * ports then flood it there as well.
      */
     if (od->n_router_ports != od->nbs->n_ports) {
         ds_put_format(&actions, "clone {outport = %s; output; }; "
-                                "outport = \""MC_FLOOD_L2"\"; output;",
+                                "outport = \""MC_UNKNOWN"\"; output;",
                       patch_op->json_key);
         ovn_lflow_add_with_hint(lflows, od, S_SWITCH_IN_L2_LKUP,
                                 priority, ds_cstr(&match),
                                 ds_cstr(&actions), stage_hint,
                                 lflow_ref);
+        if (addr_family == AF_INET) {
+            ovn_lflow_add_with_hint(lflows, od, S_SWITCH_IN_L2_LKUP,
+                                    priority + 10, ds_cstr(&m_garp),
+                                    "outport = \""MC_FLOOD_L2"\"; output;",
+                                    stage_hint, lflow_ref);
+        }
     } else {
         ds_put_format(&actions, "outport = %s; output;", patch_op->json_key);
         ovn_lflow_add_with_hint(lflows, od, S_SWITCH_IN_L2_LKUP,
@@ -8743,7 +8755,7 @@ build_lswitch_rport_arp_req_flow(
         ds_clear(&actions);
         if (od->n_router_ports != od->nbs->n_ports) {
             ds_put_format(&actions, "clone {outport = %s; output; }; "
-                                    "outport = \""MC_FLOOD_L2"\"; output;",
+                                    "outport = \""MC_UNKNOWN"\"; output;",
                           patch_op->cr_port->json_key);
             ovn_lflow_add_with_hint(lflows, od, S_SWITCH_IN_L2_LKUP,
                                     priority, ds_cstr(&match),
@@ -8762,6 +8774,7 @@ build_lswitch_rport_arp_req_flow(

     ds_destroy(&m);
     ds_destroy(&match);
+    ds_destroy(&m_garp);
     ds_destroy(&actions);
 }

@@ -10151,7 +10164,7 @@ build_lswitch_destination_lookup_bmcast(struct ovn_datapath *od,
                        "broadcast-arps-to-all-routers", true)) {
         ovn_lflow_add(lflows, od, S_SWITCH_IN_L2_LKUP, 72,
                       "eth.mcast && (arp.op == 1 || nd_ns)",
-                      "outport = \""MC_FLOOD_L2"\"; output;",
+                      "outport = \""MC_UNKNOWN"\"; output;",
                       lflow_ref);
     }

diff --git a/northd/ovn-northd.8.xml b/northd/ovn-northd.8.xml
index 480d0c151c..548d67f03e 100644
--- a/northd/ovn-northd.8.xml
+++ b/northd/ovn-northd.8.xml
@@ -2153,6 +2153,14 @@ output;
         </p>
       </li>

+      <li>
+        Priority-90 flows for each IPv4 address/VIP/NAT address owned by a
+        router port connected to the switch. These flows match GARP packets
+        for the specific IP addresses. Matched packets are forwarded to the
+        <code>MC_FLOOD_L2</code> multicast group which contains all non-router
+        logical ports.
+      </li>
+
       <li>
         Priority-90 flows for transit switches that forward registered
         IP multicast traffic to their corresponding multicast group , which
@@ -2216,8 +2224,8 @@ output;
         router port connected to the switch. These flows match ARP requests
         and ND packets for the specific IP addresses.  Matched packets are
         forwarded only to the router that owns the IP address and to the
-        <code>MC_FLOOD_L2</code> multicast group which contains all non-router
-        logical ports.
+        <code>MC_UNKNOWN</code> multicast group which contains all enabled
+        logical ports that accept unknown destination packets.
       </li>

       <li>
@@ -2240,6 +2248,6 @@ output;
       <li>
         A priority-72 flow that outputs all ARP requests and ND NS (Neighbor
         Solicitation) packets with an Ethernet broadcast or multicast
-        <code>eth.dst</code> to the <code>MC_FLOOD_L2</code> multicast group
+        <code>eth.dst</code> to the <code>MC_UNKNOWN</code> multicast group
         if <code>other_config:broadcast-arps-to-all-routers=false</code>.
       </li>
